app/kuma-dp/pkg/dataplane/readiness/component.go

```diff
@@ -0,0 +1,120 @@
+package readiness
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"os"
+	"sync/atomic"
+	"time"
+
+	"github.com/bakito/go-log-logr-adapter/adapter"
+	"github.com/pkg/errors"
+
+	"github.com/kumahq/kuma/pkg/core"
+	"github.com/kumahq/kuma/pkg/core/runtime/component"
+)
+
+const (
+	pathPrefixReady  = "/ready"
+	stateReady       = "READY"
+	stateTerminating = "TERMINATING"
+)
+
+// Reporter reports the health status of this Kuma Dataplane Proxy
+type Reporter struct {
+	socketPath    string
+	isTerminating atomic.Bool
+}
+
+var logger = core.Log.WithName("readiness")
+
+func NewReporter(socketPath string) *Reporter {
+	return &Reporter{
+		socketPath: socketPath,
+	}
+}
+
+func (r *Reporter) Start(stop <-chan struct{}) error {
+	_, err := os.Stat(r.socketPath)
+	if err == nil {
+		// File is accessible try to rename it to verify it is not open
+		newName := r.socketPath + ".bak"
+		err = os.Rename(r.socketPath, newName)
+		if err != nil {
+			return errors.Errorf("file %s exists and probably opened by another kuma-dp instance", r.socketPath)
+		}
+		err = os.Remove(newName)
+		if err != nil {
+			return errors.Errorf("not able the delete the backup file %s", newName)
+		}
+	}
+
+	lis, err := net.Listen("unix", r.socketPath)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		_ = lis.Close()
+	}()
+
+	logger.Info("starting readiness reporter",
+		"socketPath", fmt.Sprintf("unix://%s", r.socketPath),
+	)
+
+	mux := http.NewServeMux()
+	mux.HandleFunc(pathPrefixReady, r.handleReadiness)
+	server := &http.Server{
+		ReadHeaderTimeout: time.Second,
+		Handler:           mux,
+		ErrorLog:          adapter.ToStd(logger),
+	}
+
+	errCh := make(chan error)
+	go func() {
+		if err := server.Serve(lis); err != nil {
+			errCh <- err
+		}
+	}()
+
+	select {
+	case err := <-errCh:
+		return err
+	case <-stop:
+		logger.Info("stopping readiness reporter")
+		return server.Shutdown(context.Background())
+	}
+}
+
+func (r *Reporter) Terminating() {
+	r.isTerminating.Store(true)
+}
+
+func (r *Reporter) handleReadiness(writer http.ResponseWriter, req *http.Request) {
+	state := stateReady
+	stateHTTPStatus := http.StatusOK
+	if r.isTerminating.Load() {
+		state = stateTerminating
+		stateHTTPStatus = http.StatusServiceUnavailable
+	}
+
+	stateBytes := []byte(state)
+	writer.Header().Set("Content-Type", "text/plain")
+	writer.Header().Set("Content-Length", fmt.Sprintf("%d", len(stateBytes)))
+	writer.Header().Set("Cache-Control", "no-cache, max-age=0")
+	writer.Header().Set("X-Readiness-Server", "kuma-dp")
```

Comment:
I think all these should be lowercase to avoid confusing users (Envoy lowercases, kuma-dp does not)

```suggestion
	writer.Header().Set("content-type", "text/plain")
	writer.Header().Set("content-length", fmt.Sprintf("%d", len(stateBytes)))
	writer.Header().Set("cache-control", "no-cache, max-age=0")
	writer.Header().Set("x-readiness-server", "kuma-dp")
```

also aren't `x-*` headers deprecated? https://datatracker.ietf.org/doc/html/rfc6648 https://tonyxu.io/blog/http-deprecate-x-prefix/ - I'm not sure we follow that RFC but might be good to figure out.

Reply:
Right, this practice is deprecated, and it looks like many `X-*` headhers are now becoming official. One example is "x-powered-by" and it looks this one fit our scenario well. I think it's beneficial to clearly tell the client that this response is generated from `kuma-dp`.
That been said, I'm going to change it to `x-powered-by`, what do you think?

Reply:
yeah, we can do it that way 👍